File: src/services/indicators.ts
export function calculateSMA(data: number[], period: number): number[] {
  const sma: number[] = [];
  for (let i = 0; i < data.length; i++) {
    if (i < period - 1) {
      sma.push(NaN);
      continue;
    }
    let sum = 0;
    for (let j = 0; j < period; j++) {
      sum += data[i - j];
    }
    sma.push(sum / period);
  }
  return sma;
}

export function calculateEMA(data: number[], period: number): number[] {
  const ema: number[] = [];
  const multiplier = 2 / (period + 1);
  
  for (let i = 0; i < data.length; i++) {
    if (i === 0) {
      ema.push(data[0]);
    } else {
      ema.push((data[i] - ema[i - 1]) * multiplier + ema[i - 1]);
    }
  }
  return ema;
}

export function calculateRSI(data: number[], period: number = 14): number[] {
  const rsi: number[] = [];
  let gains = 0;
  let losses = 0;

  for (let i = 0; i < data.length; i++) {
    if (i === 0) {
      rsi.push(NaN);
      continue;
    }

    const diff = data[i] - data[i - 1];
    if (i <= period) {
      if (diff > 0) gains += diff;
      else losses -= diff;
      
      if (i === period) {
        const avgGain = gains / period;
        const avgLoss = losses / period;
        const rs = avgGain / (avgLoss === 0 ? 1 : avgLoss);
        rsi.push(100 - (100 / (1 + rs)));
      } else {
        rsi.push(NaN);
      }
    } else {
      const prevAvgGain = gains;
      const prevAvgLoss = losses;
      
      const currentGain = diff > 0 ? diff : 0;
      const currentLoss = diff < 0 ? -diff : 0;

      gains = (prevAvgGain * (period - 1) + currentGain) / period;
      losses = (prevAvgLoss * (period - 1) + currentLoss) / period;

      const rs = gains / (losses === 0 ? 1 : losses);
      rsi.push(100 - (100 / (1 + rs)));
    }
  }
  return rsi;
}

export function calculateMACD(data: number[], fastPeriod = 12, slowPeriod = 26, signalPeriod = 9) {
  const fastEMA = calculateEMA(data, fastPeriod);
  const slowEMA = calculateEMA(data, slowPeriod);
  
  const macdLine = fastEMA.map((fast, i) => fast - slowEMA[i]);
  
  // Calculate signal line (EMA of MACD line)
  // We need to filter out NaNs for the signal line calculation if there are any
  const validMacdLine = macdLine.filter(val => !isNaN(val));
  const signalLineValid = calculateEMA(validMacdLine, signalPeriod);
  
  const signalLine = new Array(data.length - validMacdLine.length).fill(NaN).concat(signalLineValid);
  
  const histogram = macdLine.map((macd, i) => macd - signalLine[i]);

  return { macdLine, signalLine, histogram };
}

export function calculateBollingerBands(data: number[], period = 20, multiplier = 2) {
  const sma = calculateSMA(data, period);
  const upper: number[] = [];
  const lower: number[] = [];

  for (let i = 0; i < data.length; i++) {
    if (i < period - 1) {
      upper.push(NaN);
      lower.push(NaN);
      continue;
    }

    let sumSq = 0;
    for (let j = 0; j < period; j++) {
      sumSq += Math.pow(data[i - j] - sma[i], 2);
    }
    const stdDev = Math.sqrt(sumSq / period);
    
    upper.push(sma[i] + stdDev * multiplier);
    lower.push(sma[i] - stdDev * multiplier);
  }

  return { middle: sma, upper, lower };
}

File: backend/platform_api/services/indicators.py
from math import sqrt


def calculate_sma(data: list[float], period: int) -> list[float]:
    output: list[float] = []
    for index, _ in enumerate(data):
        if index < period - 1:
            output.append(float("nan"))
            continue

        window = data[index - period + 1 : index + 1]
        output.append(sum(window) / period)
    return output


def calculate_ema(data: list[float], period: int) -> list[float]:
    output: list[float] = []
    multiplier = 2 / (period + 1)
    for index, price in enumerate(data):
        if index == 0:
            output.append(price)
            continue
        previous = output[index - 1]
        output.append((price - previous) * multiplier + previous)
    return output


def calculate_rsi(data: list[float], period: int = 14) -> list[float]:
    output = [float("nan")] * len(data)
    if len(data) <= period:
        return output

    average_gain = 0.0
    average_loss = 0.0
    for index in range(1, period + 1):
        diff = data[index] - data[index - 1]
        if diff >= 0:
            average_gain += diff
        else:
            average_loss += abs(diff)

    average_gain /= period
    average_loss /= period
    output[period] = _rsi_value(average_gain, average_loss)

    for index in range(period + 1, len(data)):
        diff = data[index] - data[index - 1]
        gain = diff if diff > 0 else 0
        loss = abs(diff) if diff < 0 else 0
        average_gain = (average_gain * (period - 1) + gain) / period
        average_loss = (average_loss * (period - 1) + loss) / period
        output[index] = _rsi_value(average_gain, average_loss)

    return output


def calculate_bollinger_bands(data: list[float], period: int = 20, multiplier: int = 2) -> tuple[list[float], list[float], list[float]]:
    middle = calculate_sma(data, period)
    upper = [float("nan")] * len(data)
    lower = [float("nan")] * len(data)

    for index in range(period - 1, len(data)):
        average = middle[index]
        window = data[index - period + 1 : index + 1]
        variance = sum((value - average) ** 2 for value in window) / period
        standard_deviation = sqrt(variance)
        upper[index] = average + standard_deviation * multiplier
        lower[index] = average - standard_deviation * multiplier

    return middle, upper, lower


def _rsi_value(average_gain: float, average_loss: float) -> float:
    if average_loss == 0:
        return 100
    relative_strength = average_gain / average_loss
    return 100 - 100 / (1 + relative_strength)
